File: content/snippets/route-support-tickets/n2.test.py
"""
These tests inject a local double instead of loading the real encoder.

What they prove: the archive is encoded once and not once per query, the
vectors are brought to length one, the cosine is computed and voted on the way
the snippet claims, the floor sends an unknown ticket to the default queue,
and ties are resolved the way the docstring says.

What they do not prove: that the encoder understands anything. The double is a
bag of words, so the very pair this rung exists for — a ticket and its
paraphrase, sharing no word — scores near nothing. That is asserted below
rather than hidden, and it is why the entry declares this snippet `stubbed`.
"""

import sys
import time
import types

import pytest

import n2
from _harness.fake_model import FakeEncoder
from n2 import DEFAULT_TEAM, MODEL_NAME, build_index, neighbours, route

# Resolved tickets, and the team that resolved each one.
ARCHIVE = [
    ("Ma facture de janvier est trop élevée", "billing"),
    ("Le prélèvement est passé deux fois ce mois-ci", "billing"),
    ("Bonjour, merci de me confirmer le remboursement de ma commande", "billing"),
    ("Impossible de me connecter depuis ce matin", "technical"),
    ("L'application plante quand j'ouvre le tableau de bord", "technical"),
    ("J'ai perdu mon mot de passe", "technical"),
    ("Mon colis n'est toujours pas arrivé", "shipping"),
    ("La livraison a été annulée par le transporteur", "shipping"),
    ("Le suivi indique livré mais je n'ai rien reçu", "shipping"),
]

TICKETS = [ticket for ticket, _ in ARCHIVE]
TEAMS = [team for _, team in ARCHIVE]

# The real encoder of this snippet returns 384 numbers per ticket; the double
# is asked for the same width, so the test exercises the real shape.
DIMENSIONS = 384

POLITENESS = "Bonjour, merci de me confirmer que vous avez bien reçu mon dossier"


def make_index(encoder=None, tickets=TICKETS, teams=TEAMS):
    return build_index(tickets, teams, encoder or FakeEncoder(DIMENSIONS))


class VecteursDonnes:
    """Un encodeur dont chaque vecteur est écrit dans le test."""

    def __init__(self, vecteurs):
        self.vecteurs = vecteurs

    def encode(self, texts):
        return [self.vecteurs[t] for t in texts]


# ---------------------------------------------------------------------------
# Point de rupture
# ---------------------------------------------------------------------------


def test_point_de_rupture_un_ticket_qui_ne_sadresse_a_personne_part_chez_facturation():
    """
    « "Bonjour, merci de me confirmer que vous avez bien reçu mon dossier" ne
    s'adresse à personne en particulier, et part chez facturation avec un score
    élevé : l'archive contient un ticket de facturation écrit avec les mêmes
    formules de politesse ».
    """
    index = make_index()
    similarity, team = neighbours(index, POLITENESS, k=1)[0]
    assert team == "billing"
    assert round(similarity, 12) == 0.617213399848
    assert route(index, POLITENESS) == "billing"
    # Le voisin est bien le ticket de facturation poli.
    vecteur = n2._unit(index["encoder"].encode([POLITENESS])[0])
    scores = [n2._dot(v, vecteur) for v in index["vectors"]]
    assert TICKETS[scores.index(max(scores))].startswith("Bonjour, merci de me confirmer")


def test_point_de_rupture_temoin_sans_ce_ticket_dans_larchive_laccident_se_deplace():
    """
    « ce dont l'archive est faite est la politique de routage » : retiré de
    l'archive, le ticket poli ne décide plus, et le même message part chez
    technique, sur deux tickets qui partagent « de me ».
    """
    gardes = [(t, e) for t, e in ARCHIVE if not t.startswith("Bonjour")]
    index = make_index(tickets=[t for t, _ in gardes], teams=[e for _, e in gardes])
    assert route(index, POLITENESS) == "technical"


def test_point_de_rupture_le_vote_tranche_de_justesse():
    """Précision du score : 0,617 pour facturation contre 0,303 + 0,303 = 0,606 pour technique."""
    found = neighbours(make_index(), POLITENESS)
    assert [team for _, team in found] == ["billing", "technical", "technical"]
    assert round(found[0][0] - (found[1][0] + found[2][0]), 3) == 0.011


# ---------------------------------------------------------------------------
# Les autres affirmations du niveau
# ---------------------------------------------------------------------------


def test_route_un_ticket_vers_lequipe_de_ses_voisins_les_plus_proches():
    assert route(make_index(), "Ma facture de février est trop élevée") == "billing"


def test_le_voisin_le_plus_proche_et_son_score():
    closest = neighbours(make_index(), "Ma facture de février est trop élevée", k=1)
    similarity, team = closest[0]
    assert team == "billing"
    assert round(similarity, 12) == 0.857142857143


@pytest.mark.xfail(strict=True, reason=(
    "INFIRMÉ : la docstring dit que « the neighbours are shown to the agent as the "
    "reason for the routing — which is more than N1's weights ever explain » ; "
    "`neighbours` rend des couples (score, équipe) sans le ticket archivé ni son "
    "rang dans l'archive : l'agent voit « billing, 0,62 », pas le ticket qui a décidé"
))
def test_infirme_les_voisins_montres_a_lagent_disent_quel_ticket_a_decide():
    found = neighbours(make_index(), POLITENESS, k=1)
    assert any(isinstance(part, str) and part in TICKETS for part in found[0])


def test_larchive_est_encodee_une_fois_et_non_a_chaque_question():
    encoder = FakeEncoder(DIMENSIONS)
    index = make_index(encoder)
    assert encoder.calls == [TICKETS]
    route(index, "Mon colis est en retard chez le transporteur")
    route(index, "Ma facture est fausse")
    assert encoder.calls[1:] == [["Mon colis est en retard chez le transporteur"], ["Ma facture est fausse"]]


def test_lindex_est_une_copie_de_larchive():
    """Docstring : « the index is the archive itself » ; regulatory : « l'index, qui en est une copie »."""
    tickets = list(TICKETS)
    index = make_index(tickets=tickets)
    tickets.pop(0)
    assert index["tickets"] == TICKETS
    assert len(index["vectors"]) == len(TICKETS)


def test_le_vote_compte_le_voisinage_et_pas_seulement_le_meilleur():
    """Docstring de route : « one archived ticket that happens to be phrased like this one is an accident, three of them are a pattern »."""
    ticket = "Mon colis est en retard chez le transporteur"
    assert [team for _, team in neighbours(make_index(), ticket)] == ["shipping", "billing", "shipping"]
    assert route(make_index(), ticket) == "shipping"
    # Témoin : réduit au meilleur voisin (k=1), le vote est le même ici ; il diffère quand le meilleur est isolé.
    vecteurs = {"q": [1.0, 0.0], "a": [0.9, 0.1], "b": [0.8, 0.2], "c": [0.8, 0.2]}
    index = build_index(["a", "b", "c"], ["x", "y", "y"], VecteursDonnes(vecteurs))
    assert route(index, "q", k=1) == "x"
    assert route(index, "q", k=3) == "y"


def test_a_egalite_de_score_le_tri_garde_lordre_de_larchive():
    """Commentaire : « A stable sort, so two equally close tickets always come back in archive order »."""
    vecteurs = {"q": [1.0, 0.0], "a": [0.6, 0.8], "b": [0.6, 0.8], "c": [0.6, 0.8]}
    index = build_index(["a", "b", "c"], ["x", "y", "z"], VecteursDonnes(vecteurs))
    assert [team for _, team in neighbours(index, "q")] == ["x", "y", "z"]


def test_a_egalite_de_vote_lequipe_du_plus_proche_lemporte():
    """Commentaire : « Ties go to the team of the closest neighbour, which is the first key inserted above »."""
    # Composantes dyadiques, normes exactement 1 : votes x = 0,5 ; y = 0,25 + 0,25 = 0,5, égalité exacte.
    vecteurs = {"q": [1.0, 0, 0, 0, 0], "a": [0.5, 0.5, 0.5, 0.5, 0],
                "b": [0.25, 0.75, 0.5, 0.25, 0.25], "c": [0.25, 0.25, 0.25, 0.5, 0.75]}
    index = build_index(["a", "b", "c"], ["x", "y", "y"], VecteursDonnes(vecteurs))
    assert route(index, "q") == "x"
    assert [round(sc, 12) for sc, _ in neighbours(index, "q")] == [0.5, 0.25, 0.25]
    # Témoin : si le plus proche est de l'équipe y, c'est y qui l'emporte à égalité.
    index = build_index(["a", "b", "c"], ["y", "x", "x"], VecteursDonnes(vecteurs))
    assert route(index, "q") == "y"


def test_le_plancher_juste_au_dessus_et_juste_en_dessous():
    """Docstring de route : « The floor is what keeps the default queue of N0 alive »."""
    vecteurs = {"q": [1.0, 0.0], "a": [0.25, 0.9682458365518543]}
    index = build_index(["a"], ["x"], VecteursDonnes(vecteurs))
    assert route(index, "q", k=1) == "x"
    assert route(index, "q", k=1, min_similarity=0.2500001) == DEFAULT_TEAM


def test_un_ticket_que_larchive_na_jamais_vu_part_dans_la_file_par_defaut():
    ticket = "Votre entrepôt accepte-t-il les visites scolaires"
    assert [similarity for similarity, _ in neighbours(make_index(), ticket)] == [0.0, 0.0, 0.0]
    assert route(make_index(), ticket) == DEFAULT_TEAM


def test_ce_que_le_double_ne_peut_pas_prouver():
    """
    Docstring : « "je n'arrive plus à entrer dans mon espace" lands next to the
    archived tickets about a lost password » — non démontrable avec le double,
    qui est un sac de mots : il met le ticket du mot de passe troisième.
    """
    index = make_index()
    ticket = "Je n'arrive plus à entrer dans mon espace client"
    assert [team for _, team in neighbours(index, ticket)] == ["shipping", "shipping", "technical"]
    assert route(index, ticket) == DEFAULT_TEAM


def test_lencodeur_par_defaut_a_la_forme_de_sentence_transformers(monkeypatch):
    """
    Docstring de load_encoder : « fetched once, then held in memory and run
    locally ». Double à la forme publiée : `SentenceTransformer(name)` puis
    `.encode(list_of_str)` qui rend un tableau (liste de vecteurs).
    """
    charges = []

    class SentenceTransformer:
        def __init__(self, name):
            charges.append(name)
            self.inner = FakeEncoder(DIMENSIONS)

        def encode(self, sentences):
            return [list(v) for v in self.inner.encode(sentences)]

    module = types.ModuleType("sentence_transformers")
    module.SentenceTransformer = SentenceTransformer
    monkeypatch.setitem(sys.modules, "sentence_transformers", module)
    index = build_index(TICKETS, TEAMS)
    assert route(index, "Ma facture de février est trop élevée") == "billing"
    route(index, "Mon colis est perdu")
    assert charges == [MODEL_NAME]


def test_deterministe():
    assert neighbours(make_index(), POLITENESS) == neighbours(make_index(), POLITENESS)


# ---------------------------------------------------------------------------
# Cas de production
# ---------------------------------------------------------------------------


def test_production_ticket_vide_archive_vide_et_k_nul():
    assert route(make_index(), "") == DEFAULT_TEAM
    assert route(make_index(tickets=[], teams=[]), "Ma facture") == DEFAULT_TEAM
    assert route(make_index(), "Ma facture de janvier est trop élevée", k=0) == DEFAULT_TEAM
    assert len(neighbours(make_index(), "Ma facture", k=50)) == len(TICKETS)


def test_production_nfd_emoji_bom_et_casse_arrivent_a_lencodeur_tels_quels():
    encoder = FakeEncoder(DIMENSIONS)
    index = make_index(encoder)
    ticket = "\ufeff📦 COLIS perdu, cafe\u0301\u00a0!"
    route(index, ticket)
    assert encoder.calls[-1] == [ticket]


def test_production_une_archive_de_900_tickets_repond_vite():
    tickets = [f"{t} numero {i}" for i in range(100) for t, _ in ARCHIVE]
    teams = [e for _ in range(100) for _, e in ARCHIVE]
    index = make_index(tickets=tickets, teams=teams)
    debut = time.perf_counter()
    assert route(index, "Ma facture de février est trop élevée") == "billing"
    assert time.perf_counter() - debut < 5.0


@pytest.mark.xfail(strict=True, reason=(
    "DÉFAUT : la docstring prévient que l'archive doit être ré-encodée quand le "
    "modèle change, mais rien ne le vérifie. Un index de vecteurs de 384 nombres "
    "interrogé par un encodeur de 768 n'est pas refusé : Python tronque le produit "
    "scalaire (`zip`) et route sur un score faux mais plausible (facturation, "
    "0,38) ; JavaScript tronque de même, et calcule NaN quand l'encodeur est plus "
    "étroit que l'index, ce qui envoie tout en file par défaut"
))
def test_defaut_un_index_encode_par_un_autre_modele_est_refuse():
    index = make_index()
    index["encoder"] = FakeEncoder(768)
    with pytest.raises(ValueError):
        route(index, "Ma facture de janvier")

File: content/snippets/route-support-tickets/n2.test.js
/**
 * These tests inject a local double instead of loading the real encoder.
 *
 * What they prove: the archive is encoded once and not once per query, the
 * vectors are brought to length one, the cosine is computed and voted on the
 * way the snippet claims, the floor sends an unknown ticket to the default
 * queue, and ties are resolved the way the docstring says.
 *
 * What they do not prove: that the encoder understands anything. The double is
 * a bag of words, so the very pair this rung exists for — a ticket and its
 * paraphrase, sharing no word — scores near nothing. That is asserted below
 * rather than hidden, and it is why the entry declares this snippet `stubbed`.
 */
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { register } from 'node:module';
import { performance } from 'node:perf_hooks';
import { FakeEncoder } from '../_harness/fake-model.mjs';
import { DEFAULT_TEAM, MODEL_NAME, buildIndex, neighbours, route } from './n2.js';

// Resolved tickets, and the team that resolved each one.
const ARCHIVE = [
  ['Ma facture de janvier est trop élevée', 'billing'],
  ['Le prélèvement est passé deux fois ce mois-ci', 'billing'],
  ['Bonjour, merci de me confirmer le remboursement de ma commande', 'billing'],
  ['Impossible de me connecter depuis ce matin', 'technical'],
  ["L'application plante quand j'ouvre le tableau de bord", 'technical'],
  ["J'ai perdu mon mot de passe", 'technical'],
  ["Mon colis n'est toujours pas arrivé", 'shipping'],
  ['La livraison a été annulée par le transporteur', 'shipping'],
  ["Le suivi indique livré mais je n'ai rien reçu", 'shipping'],
];

const TICKETS = ARCHIVE.map(([ticket]) => ticket);
const TEAMS = ARCHIVE.map(([, team]) => team);

// The real encoder of this snippet returns 384 numbers per ticket; the double
// is asked for the same width, so the test exercises the real shape.
const DIMENSIONS = 384;

const POLITENESS = 'Bonjour, merci de me confirmer que vous avez bien reçu mon dossier';

const makeIndex = (encoder, tickets = TICKETS, teams = TEAMS) =>
  buildIndex(tickets, teams, encoder ?? new FakeEncoder(DIMENSIONS));

/** Un encodeur dont chaque vecteur est écrit dans le test. */
const vecteursDonnes = (vecteurs) => ({ encode: async (texts) => texts.map((t) => vecteurs[t]) });

/*
 * Un faux module `@huggingface/transformers` à la forme publiée :
 * `pipeline(task, model, options)` rend une fonction `extract(texts, { pooling })`
 * dont le résultat a `.tolist()`. Servi par un crochet de résolution, le paquet
 * n'étant pas installé.
 */
const FAUX_TRANSFORMERS = `
const words = (t) => t.toLowerCase().replace(/[^\\p{L}\\p{N}]+/gu, ' ').split(' ').filter(Boolean);
const hash = (w) => { let h = 2166136261; for (const c of w) h = Math.imul(h ^ c.codePointAt(0), 16777619) >>> 0; return h; };
export async function pipeline(task, model, options) {
  globalThis.__transformers.loads.push([task, model, options]);
  return async (texts, options) => {
    globalThis.__transformers.calls.push(options);
    return { tolist: () => texts.map((t) => { const v = new Array(384).fill(0); for (const w of words(t)) v[hash(w) % 384] += 1; return v; }) };
  };
}`;
globalThis.__transformers = { loads: [], calls: [] };
register(`data:text/javascript,${encodeURIComponent(`
export async function resolve(specifier, context, next) {
  if (specifier === '@huggingface/transformers') {
    return { url: 'data:text/javascript,' + ${JSON.stringify(encodeURIComponent(FAUX_TRANSFORMERS))}, shortCircuit: true };
  }
  return next(specifier, context);
}`)}`);

// ---------------------------------------------------------------------------
// Point de rupture
// ---------------------------------------------------------------------------

test('point de rupture : un ticket qui ne s’adresse à personne part chez facturation', async () => {
  const index = await makeIndex();
  const [[similarity, team]] = await neighbours(index, POLITENESS, 1);
  assert.equal(team, 'billing');
  assert.equal(Number(similarity.toFixed(12)), 0.617213399848);
  assert.equal(await route(index, POLITENESS), 'billing');
  // Le voisin est bien le ticket de facturation poli : seul, il rend le même score.
  const seul = await makeIndex(undefined, [TICKETS[2]], [TEAMS[2]]);
  assert.equal(Number((await neighbours(seul, POLITENESS, 1))[0][0].toFixed(12)), 0.617213399848);
});

test('point de rupture : témoin, sans ce ticket dans l’archive l’accident se déplace', async () => {
  const gardes = ARCHIVE.filter(([t]) => !t.startsWith('Bonjour'));
  const index = await makeIndex(undefined, gardes.map(([t]) => t), gardes.map(([, e]) => e));
  assert.equal(await route(index, POLITENESS), 'technical');
});

test('point de rupture : le vote tranche de justesse', async () => {
  const found = await neighbours(await makeIndex(), POLITENESS);
  assert.deepEqual(found.map(([, team]) => team), ['billing', 'technical', 'technical']);
  assert.equal(Number((found[0][0] - (found[1][0] + found[2][0])).toFixed(3)), 0.011);
});

// ---------------------------------------------------------------------------
// Les autres affirmations du niveau
// ---------------------------------------------------------------------------

test('route un ticket vers l’équipe de ses voisins les plus proches', async () => {
  assert.equal(await route(await makeIndex(), 'Ma facture de février est trop élevée'), 'billing');
});

test('le voisin le plus proche et son score', async () => {
  const [[similarity, team]] = await neighbours(await makeIndex(), 'Ma facture de février est trop élevée', 1);
  assert.equal(team, 'billing');
  assert.equal(Number(similarity.toFixed(12)), 0.857142857143);
});

test('les voisins rendus sont un score et une équipe, pas le ticket qui a décidé', async () => {
  // « The k nearest resolved tickets, best first, with their cosine score » :
  // l'agent reçoit « billing, 0,62 », pas le ticket archivé qui a décidé, et la
  // fiche ne dit pas le contraire.
  const [premier] = await neighbours(await makeIndex(), POLITENESS, 1);
  const [score, team] = premier;
  assert.equal(typeof score, 'number');
  assert.ok(TEAMS.includes(team));
  assert.ok(!premier.some((part) => typeof part === 'string' && TICKETS.includes(part)));
});

test('l’archive est encodée une fois et non à chaque question', async () => {
  const encoder = new FakeEncoder(DIMENSIONS);
  const index = await makeIndex(encoder);
  assert.deepEqual(encoder.calls, [TICKETS]);
  await route(index, 'Mon colis est en retard chez le transporteur');
  await route(index, 'Ma facture est fausse');
  assert.deepEqual(encoder.calls.slice(1), [['Mon colis est en retard chez le transporteur'], ['Ma facture est fausse']]);
});

test('l’index est une copie de l’archive', async () => {
  const tickets = [...TICKETS];
  const index = await makeIndex(undefined, tickets);
  tickets.shift();
  assert.deepEqual(index.tickets, TICKETS);
  assert.equal(index.vectors.length, TICKETS.length);
});

test('le vote compte le voisinage, et pas seulement le meilleur', async () => {
  const ticket = 'Mon colis est en retard chez le transporteur';
  assert.deepEqual((await neighbours(await makeIndex(), ticket)).map(([, t]) => t), ['shipping', 'billing', 'shipping']);
  assert.equal(await route(await makeIndex(), ticket), 'shipping');
  const vecteurs = { q: [1, 0], a: [0.9, 0.1], b: [0.8, 0.2], c: [0.8, 0.2] };
  const index = await buildIndex(['a', 'b', 'c'], ['x', 'y', 'y'], vecteursDonnes(vecteurs));
  assert.equal(await route(index, 'q', { k: 1 }), 'x');
  assert.equal(await route(index, 'q', { k: 3 }), 'y');
});

test('à égalité de score, le tri garde l’ordre de l’archive', async () => {
  const vecteurs = { q: [1, 0], a: [0.6, 0.8], b: [0.6, 0.8], c: [0.6, 0.8] };
  const index = await buildIndex(['a', 'b', 'c'], ['x', 'y', 'z'], vecteursDonnes(vecteurs));
  assert.deepEqual((await neighbours(index, 'q')).map(([, t]) => t), ['x', 'y', 'z']);
});

test('à égalité de vote, l’équipe du plus proche l’emporte', async () => {
  const vecteurs = { q: [1, 0, 0, 0, 0], a: [0.5, 0.5, 0.5, 0.5, 0], b: [0.25, 0.75, 0.5, 0.25, 0.25], c: [0.25, 0.25, 0.25, 0.5, 0.75] };
  let index = await buildIndex(['a', 'b', 'c'], ['x', 'y', 'y'], vecteursDonnes(vecteurs));
  assert.equal(await route(index, 'q'), 'x');
  assert.deepEqual((await neighbours(index, 'q')).map(([s]) => s), [0.5, 0.25, 0.25]);
  index = await buildIndex(['a', 'b', 'c'], ['y', 'x', 'x'], vecteursDonnes(vecteurs));
  assert.equal(await route(index, 'q'), 'y');
});

test('le plancher juste au-dessus et juste en dessous', async () => {
  const vecteurs = { q: [1, 0], a: [0.25, 0.9682458365518543] };
  const index = await buildIndex(['a'], ['x'], vecteursDonnes(vecteurs));
  assert.equal(await route(index, 'q', { k: 1 }), 'x');
  assert.equal(await route(index, 'q', { k: 1, minSimilarity: 0.2500001 }), DEFAULT_TEAM);
});

test('un ticket que l’archive n’a jamais vu part dans la file par défaut', async () => {
  const ticket = 'Votre entrepôt accepte-t-il les visites scolaires';
  assert.deepEqual((await neighbours(await makeIndex(), ticket)).map(([s]) => s), [0, 0, 0]);
  assert.equal(await route(await makeIndex(), ticket), DEFAULT_TEAM);
});

test('ce que le double ne peut pas prouver', async () => {
  const index = await makeIndex();
  const ticket = "Je n'arrive plus à entrer dans mon espace client";
  assert.deepEqual((await neighbours(index, ticket)).map(([, team]) => team), ['shipping', 'shipping', 'technical']);
  assert.equal(await route(index, ticket), DEFAULT_TEAM);
});

test('l’encodeur par défaut a la forme de @huggingface/transformers', async () => {
  // `pipeline('feature-extraction', name, { dtype })` puis
  // `extract(texts, { pooling: 'mean' }).tolist()`.
  globalThis.__transformers = { loads: [], calls: [] };
  const index = await buildIndex(TICKETS, TEAMS);
  assert.equal(await route(index, 'Ma facture de février est trop élevée'), 'billing');
  await route(index, 'Mon colis est perdu');
  assert.deepEqual(globalThis.__transformers.loads, [['feature-extraction', MODEL_NAME, { dtype: 'fp32' }]]);
  assert.deepEqual(globalThis.__transformers.calls[0], { pooling: 'mean' });
});

test('déterministe', async () => {
  assert.deepEqual(await neighbours(await makeIndex(), POLITENESS), await neighbours(await makeIndex(), POLITENESS));
});

// ---------------------------------------------------------------------------
// Cas de production
// ---------------------------------------------------------------------------

test('production : ticket vide, archive vide et k nul', async () => {
  assert.equal(await route(await makeIndex(), ''), DEFAULT_TEAM);
  assert.equal(await route(await makeIndex(undefined, [], []), 'Ma facture'), DEFAULT_TEAM);
  assert.equal(await route(await makeIndex(), 'Ma facture de janvier est trop élevée', { k: 0 }), DEFAULT_TEAM);
  assert.equal((await neighbours(await makeIndex(), 'Ma facture', 50)).length, TICKETS.length);
});

test('production : NFD, emoji, BOM et casse arrivent à l’encodeur tels quels', async () => {
  const encoder = new FakeEncoder(DIMENSIONS);
  const index = await makeIndex(encoder);
  const ticket = '\ufeff📦 COLIS perdu, cafe\u0301\u00a0!';
  await route(index, ticket);
  assert.deepEqual(encoder.calls.at(-1), [ticket]);
});

test('production : une archive de 900 tickets répond vite', async () => {
  const tickets = [];
  const teams = [];
  for (let i = 0; i < 100; i += 1) for (const [t, e] of ARCHIVE) { tickets.push(`${t} numero ${i}`); teams.push(e); }
  const index = await makeIndex(undefined, tickets, teams);
  const debut = performance.now();
  assert.equal(await route(index, 'Ma facture de février est trop élevée'), 'billing');
  assert.ok(performance.now() - debut < 5000);
});

test('un index encodé par un autre modèle est refusé', async () => {
  // La largeur du vecteur est vérifiée, dans les deux sens : un encodeur plus
  // large que l'index (768 contre 384) comme un encodeur plus étroit rendraient
  // un score faux, l'un tronqué, l'autre plein de NaN.
  const refus = { name: 'RangeError', message: /encoded by another model/ };
  let index = await makeIndex();
  index.encoder = new FakeEncoder(768);
  await assert.rejects(() => route(index, 'Ma facture de janvier'), refus);
  index = await makeIndex(new FakeEncoder(768));
  index.encoder = new FakeEncoder(DIMENSIONS);
  await assert.rejects(() => route(index, 'Ma facture de janvier'), refus);
});
